### PR TITLE
plot legend merge matching lines (#1616)

### PR DESCRIPTION
**Pull request by mnavasiolava**
Legend lines will be merged if the following parameters are match: displayName, color and outlineColor.
If there are two identical lines with LOD they will be merged and LOD hint in legend will manage them both (if lines have different lodTypes (box,avg,river) the 'mixed' type will be displayed in legend).
If there are two identical lines: one with LOD and another without LOD, they will be merged and LOD hint will be displayed (in this case it'll manage only one line).

For test you can use https://www.dropbox.com/s/merkf04vuvqmvy5/1616.bkr?dl=0

### CODE REVIEW

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js [hunk 1]
```diff
@@ -533,6 +533,70 @@
...
         };
 
+        scope.prepareMergedLegendData = function() {
+          var data = scope.stdmodel.data;
+
+          var mergedLines = {};
+          var lineUniqueAttributesSet = {};
+
+          function getColorInfoUid(dat) {
+            var color = plotUtils.createColor(dat.color, dat.color_opacity),
```
Comment (samccone) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:543]: multiline comma separated var defs will not pass our jscs config.

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js [hunk 2]
```diff
@@ -652,87 +718,138 @@
...
-            
-            for (var i = 0; i < data.length; i++) {
-              data[i].showItem = scope.showAllItems;
-              if (data[i].showItem === false) {
-                data[i].clearTips(scope);
-                if (data[i].isLodItem === true) {
-                  data[i].lodOn = false;
-                  scope.setLodHint(data[i]);
+
+            for (var lineId in scope.legendMergedLines) {
```
Comment (samccone) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:807]: for in loops can be dangerous without the hasOwnProperty check

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js [hunk 3]
```diff
@@ -565,7 +629,9 @@
           }
           legend.css(scope.legendPosition);
 
-          if (scope.legendableItem > 1) {  // skip "All" check when there is only one line
+          scope.legendMergedLines = scope.prepareMergedLegendData();
+
+          if (Object.keys(scope.legendMergedLines).length > 1) {  // skip "All" check when there is only one line
```
Comment (samccone) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:634]: This is going to fail on fairly recent versions of safari, lets use _.keys

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js (hunk 2 above)
Comment (mnavasiolava) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:807]: but there is the hasOwnProperty check (next line)

Comment (samccone) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:807]: whoops nm :palm_tree:  missed that :)

Comment (mnavasiolava) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:807]: :)

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js (hunk 1 above)
Comment (mnavasiolava) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:543]: i borrowed this style from here - bko-plot.js L674, 675

File: core/src/main/web/outputdisplay/bko-plot/bko-plot.js (hunk 3 above)
Comment (scottdraves) [inline core/src/main/web/outputdisplay/bko-plot/bko-plot.js:634]: What's the problem on Safari?

**Comment by scottdraves** [comment]
:+1:

**Pull request merged**